### PR TITLE
RPE-715: refactoring MessageProcessor into EnvelopeEventProcessor: IMessageHandler

### PR DESCRIPTION
**Pull request by KevinJCross**
JIRA link (if applicable)
https://tools.hmcts.net/jira/browse/RPE-715

Change description
Extracted out the Envelope processor part out into new class and rename the current EnvelopeProcessor to the Message Processor to indicate the message layer processing ( scheduled loop)
Also refactored to the Azure API IMessageHandler to help with future refactoring to use the Azure sdk for the Azure ServiceBus.
Note: Im not too sure about all the Completable futures used. Id personally prefer using RxJava for the Async bits but this is a quick first working draft.

Does this PR introduce a breaking change? (check one with "x")

[ ] Yes
[x ] No

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        given(someMessage.getMessageId()).willReturn(MSG_ID);
+
+        // when
+        envelopeEventProcessor.onMessageAsync(someMessage).get();
+
+        // then
+        ArgumentCaptor<String> argumentCaptor = ArgumentCaptor.forClass(String.class);
+        verify(bulkScanProcessorClient).getEnvelopeById(argumentCaptor.capture());
+        assertThat(argumentCaptor.getValue()).isEqualTo(MSG_ID);
+    }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java]: a test for what happens if `bulkScanProcessorClient` throws an exception would be nice

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java [hunk 2]
```diff
@@ -22,47 +18,44 @@
...
-    ) {
+    public MessageProcessor(ReceiverProvider receiverProvider, IMessageHandler envelopeProcessor) {
         this.receiverProvider = receiverProvider;
-        this.bulkScanProcessorClient = bulkScanProcessorClient;
+        this.envelopeProcessor = envelopeProcessor;
     }
 
     @Scheduled(fixedDelayString = "${queue.read-interval}")
     public void run() {
-        logger.info("Checking queue for new messages");
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java]: is this related?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java [hunk 3]
```diff
@@ -22,47 +18,44 @@
...
             IMessageReceiver msgReceiver = receiverProvider.get();
-            IMessage msg = msgReceiver.receive();
-            while (msg != null) {
-                handle(msgReceiver, msg);
-                msg = msgReceiver.receive();
+            while ((msg = msgReceiver.receive()) != null) {
+                process(msg);
+                msgReceiver.complete(msg.getLockToken());
             }
-        } catch (ServiceBusException | ConnectionException exc) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java]: are those no longer thrown?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java [hunk 4]
```diff
@@ -2,81 +2,93 @@
...
+    @Mock
+    private IMessageReceiver receiver;
+    @Mock
+    private EnvelopeEventProcessor envelopeEventProcessor;
 
-    @Mock private BulkScanProcessorClient bulkScanProcessorClient;
-    @Mock private ReceiverProvider receiverProvider;
-
-    @Mock private IMessage someMessage;
-    @Mock private IMessageReceiver receiver;
```
[the remark this answers is not in the document]
Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java]: Ok this was due to the default intelliJ setting.
So is it the standard to have the Annotation on the same line ?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Test
+    public void should_not_do_anything_in_notify() {
+        // when
+        envelopeEventProcessor.notifyException(null, null);
```
[the remark this answers is not in the document]
Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java]: Its for the code coverage.
currently the method does nothing. I future it could be handling some specialised service buss exceptions.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+public class EnvelopeEventProcessorTest {
+
+    @Mock
+    BulkScanProcessorClient bulkScanProcessorClient;
+
+    private EnvelopeEventProcessor envelopeEventProcessor;
+
+    @Mock
+    private IMessage someMessage;
+    private static final String MSG_ID = "hello!";
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java]: Weird order: mock, field, mock, constant.
But if checkstyle doesn't complain, I guess it's fine ;)

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java (hunk 2 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java]: It  will fill the log with "Checking queue for new messages" which is not particularly helpful. 
Maybe putting it as a debug message or removing it (I took this option) would be the most sensible.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java (hunk 3 above)
Comment (KevinJCross) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessor.java]: they are but the action taken is the same. 
Adding the exception to the log (as is now) will provide the information needed.
We collapsed this due the the code coverage checks failing.

Since the action is the same it seems sensible to simplify the code.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java (hunk 1 above)
Comment (KevinJCross) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java]: Sure ... Ill add the try catch that I commented on too.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java]: There's probably a mixture of both across the projects (for `@Mock`).

[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java]: My point is: keep changes in single PR as limited as possible to make it easier and faster review and keep number of discussed topics to minimal as well :)

[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/MessageProcessorTest.java]: Plus we normally squash changes -> commit messages should describe what the change was about

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessor.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+         * NOTE: this is done here instead of offloading to the forkJoin pool "CompletableFuture.runAsync()"
+         * because we probably should think about a threading model before doing this.
+         * Maybe consider using Netflix's RxJava too (much simpler than CompletableFuture).
+         */
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        try {
+            process(message);
+            completableFuture.complete(null);
+        } catch (Throwable t) {
+            completableFuture.completeExceptionally(t);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessor.java]: missing logging?

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+public class EnvelopeEventProcessorTest {
+
+    @Mock
+    BulkScanProcessorClient bulkScanProcessorClient;
+
+    private EnvelopeEventProcessor processor;
+
+    @Mock
+    private IMessage someMessage;
+    private static final String MSG_ID = "hello!";
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessorTest.java]: would be nice to put those 4 fields in some order. Extra points for consistent new lines and access midifiers ;)
